File: src/test/java/no/nav/sosialhjelp/soknad/consumer/arbeidsforhold/ArbeidsforholdServiceTest.java
package no.nav.sosialhjelp.soknad.consumer.arbeidsforhold;

import no.nav.sosialhjelp.soknad.consumer.arbeidsforhold.dto.AnsettelsesperiodeDto;
import no.nav.sosialhjelp.soknad.consumer.arbeidsforhold.dto.ArbeidsavtaleDto;
import no.nav.sosialhjelp.soknad.consumer.arbeidsforhold.dto.ArbeidsforholdDto;
import no.nav.sosialhjelp.soknad.consumer.arbeidsforhold.dto.OrganisasjonDto;
import no.nav.sosialhjelp.soknad.consumer.arbeidsforhold.dto.PeriodeDto;
import no.nav.sosialhjelp.soknad.consumer.arbeidsforhold.dto.PersonDto;
import no.nav.sosialhjelp.soknad.domain.model.Arbeidsforhold;
import no.nav.sosialhjelp.soknad.organisasjon.OrganisasjonService;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.LocalDate;
import java.util.List;
import java.util.stream.Collectors;

import static java.time.format.DateTimeFormatter.ISO_LOCAL_DATE;
import static java.util.Arrays.stream;
import static java.util.Collections.singletonList;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class ArbeidsforholdServiceTest {

    @Mock
    private ArbeidsforholdConsumer arbeidsforholdConsumer;

    @Mock
    private OrganisasjonService organisasjonService;

    @InjectMocks
    private ArbeidsforholdService service;

    private String fnr = "11111111111";
    private String orgnr = "orgnr";
    private String orgNavn = "Testbedriften A/S";
    private LocalDate fom = LocalDate.now().minusMonths(1);
    private LocalDate tom = LocalDate.now();

    @Test
    void skalMappeDtoTilArbeidsforhold() {
        when(arbeidsforholdConsumer.finnArbeidsforholdForArbeidstaker(fnr)).thenReturn(singletonList(createArbeidsforhold(true, fom, tom)));
        when(organisasjonService.hentOrgNavn(anyString())).thenReturn(orgNavn);

        List<Arbeidsforhold> arbeidsforholdList = service.hentArbeidsforhold(fnr);
        Arbeidsforhold arbeidsforhold = arbeidsforholdList.get(0);

        assertThat(arbeidsforhold.arbeidsgivernavn).isEqualTo(orgNavn);
        assertThat(arbeidsforhold.harFastStilling).isTrue();
        assertThat(arbeidsforhold.fastStillingsprosent.longValue()).isEqualTo(100L);
        assertThat(arbeidsforhold.orgnr).isEqualTo(orgnr);
        assertThat(arbeidsforhold.fom).isEqualTo(fom.format(ISO_LOCAL_DATE));
        assertThat(arbeidsforhold.tom).isEqualTo(tom.format(ISO_LOCAL_DATE));
        assertThat(arbeidsforhold.edagId.longValue()).isEqualTo(1337L);
    }

    @Test
    void skalSetteArbeidsgivernavnTilOrgnrHvisArbeidsgiverErOrganisasjon() {
        when(arbeidsforholdConsumer.finnArbeidsforholdForArbeidstaker(fnr)).thenReturn(singletonList(createArbeidsforhold(false, fom, tom)));

        List<Arbeidsforhold> arbeidsforholdList = service.hentArbeidsforhold(fnr);
        Arbeidsforhold arbeidsforhold = arbeidsforholdList.get(0);

        assertThat(arbeidsforhold.arbeidsgivernavn).isEqualTo("Privatperson");
        assertThat(arbeidsforhold.orgnr).isNull();
    }

    @Test
    void skalAddereStillingsprosentFraArbeidsavtaler() {
        when(arbeidsforholdConsumer.finnArbeidsforholdForArbeidstaker(fnr))
                .thenReturn(singletonList(createArbeidsforholdMedFlereArbeidsavtaler(12.3, 45.6)));
        when(organisasjonService.hentOrgNavn(anyString())).thenReturn(orgNavn);

        List<Arbeidsforhold> arbeidsforholdList = service.hentArbeidsforhold(fnr);
        Arbeidsforhold arbeidsforhold = arbeidsforholdList.get(0);

        // desimaler strippes fra double til long
        assertThat(arbeidsforhold.fastStillingsprosent.longValue()).isEqualTo(57L);
    }

    @Test
    void ansettelsesperiodeTomKanVæreNull() {
        when(arbeidsforholdConsumer.finnArbeidsforholdForArbeidstaker(fnr)).thenReturn(singletonList(createArbeidsforhold(true, fom, null)));
        when(organisasjonService.hentOrgNavn(anyString())).thenReturn(orgNavn);

        List<Arbeidsforhold> arbeidsforholdList = service.hentArbeidsforhold(fnr);
        Arbeidsforhold arbeidsforhold = arbeidsforholdList.get(0);

        assertThat(arbeidsforhold.arbeidsgivernavn).isEqualTo(orgNavn);
        assertThat(arbeidsforhold.harFastStilling).isTrue();
        assertThat(arbeidsforhold.fastStillingsprosent.longValue()).isEqualTo(100L);
        assertThat(arbeidsforhold.orgnr).isEqualTo(orgnr);
        assertThat(arbeidsforhold.fom).isEqualTo(fom.format(ISO_LOCAL_DATE));
        assertThat(arbeidsforhold.tom).isNull();
        assertThat(arbeidsforhold.edagId.longValue()).isEqualTo(1337L);
    }

    private ArbeidsforholdDto createArbeidsforhold(boolean erArbeidsgiverOrganisasjon, LocalDate fom, LocalDate tom) {
        AnsettelsesperiodeDto ansettelsesperiodeDto = new AnsettelsesperiodeDto(new PeriodeDto(fom, tom));
        ArbeidsavtaleDto arbeidsavtaleDto = createArbeidsavtale(100.0);
        return new ArbeidsforholdDto(
                ansettelsesperiodeDto,
                singletonList(arbeidsavtaleDto),
                "arbeidsforholdId",
                erArbeidsgiverOrganisasjon ? createArbeidsgiverOrganisasjon() : createArbeidsgiverPerson(),
                createArbeidstaker(),
                1337L);
    }

    private ArbeidsforholdDto createArbeidsforholdMedFlereArbeidsavtaler(double... stillingsprosenter) {
        AnsettelsesperiodeDto ansettelsesperiodeDto = new AnsettelsesperiodeDto(new PeriodeDto(fom, tom));
        List<ArbeidsavtaleDto> arbeidsavtaler = stream(stillingsprosenter)
                .boxed()
                .map(this::createArbeidsavtale)
                .collect(Collectors.toList());
        return new ArbeidsforholdDto(
                ansettelsesperiodeDto,
                arbeidsavtaler,
                "arbeidsforholdId",
                createArbeidsgiverOrganisasjon(),
                createArbeidstaker(),
                1337L);
    }

    private ArbeidsavtaleDto createArbeidsavtale(double stillingsprosent) {
        return new ArbeidsavtaleDto(stillingsprosent);
    }

    private OrganisasjonDto createArbeidsgiverOrganisasjon() {
        return new OrganisasjonDto(orgnr, "Organisasjon");
    }

    private PersonDto createArbeidsgiverPerson() {
        return new PersonDto("arbeidsgiver_fnr", "aktoerid", "Person");
    }

    private PersonDto createArbeidstaker() {
        return new PersonDto("arbeidstaker_fnr", "aktoerid", "Person");
    }
}

File: src/test/kotlin/no/nav/sosialhjelp/soknad/arbeid/ArbeidsforholdServiceTest.kt
package no.nav.sosialhjelp.soknad.arbeid

import io.getunleash.Unleash
import io.mockk.every
import io.mockk.mockk
import no.nav.sosialhjelp.soknad.arbeid.domain.Arbeidsforhold
import no.nav.sosialhjelp.soknad.arbeid.dto.AnsettelsesperiodeDto
import no.nav.sosialhjelp.soknad.arbeid.dto.ArbeidsavtaleDto
import no.nav.sosialhjelp.soknad.arbeid.dto.ArbeidsforholdDto
import no.nav.sosialhjelp.soknad.arbeid.dto.OrganisasjonDto
import no.nav.sosialhjelp.soknad.arbeid.dto.PeriodeDto
import no.nav.sosialhjelp.soknad.arbeid.dto.PersonDto
import no.nav.sosialhjelp.soknad.organisasjon.OrganisasjonService
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.time.LocalDate
import java.time.format.DateTimeFormatter

internal class ArbeidsforholdServiceTest {

    private val aaregClient: AaregClient = mockk()
    private val organisasjonService: OrganisasjonService = mockk()
    private val unleash: Unleash = mockk()
    private val arbeidsforholdService = ArbeidsforholdService(aaregClient, organisasjonService)

    private val fnr = "11111111111"
    private val orgnr = "orgnr"
    private val orgNavn = "Testbedriften A/S"
    private val fom = LocalDate.now().minusMonths(1)
    private val tom = LocalDate.now()

    private val arbeidsgiverOrganisasjon = OrganisasjonDto(organisasjonsnummer = orgnr, type = "Organisasjon")
    private val arbeidsgiverPerson = PersonDto(
        offentligIdent = "arbeidsgiver_fnr",
        aktoerId = "aktoerid",
        type = "Person"
    )
    private val arbeidstaker = PersonDto(offentligIdent = "arbeidstaker_fnr", aktoerId = "aktoerid", type = "Person")

    @BeforeEach
    fun setUp() {
        every { unleash.isEnabled(any(), false) } returns false
    }

    @Test
    internal fun skalMappeDtoTilArbeidsforhold() {
        every { aaregClient.finnArbeidsforholdForArbeidstaker(fnr) } returns listOf(createArbeidsforhold(true, fom, tom))
        every { organisasjonService.hentOrgNavn(any()) } returns orgNavn

        val arbeidsforholdList: List<Arbeidsforhold>? = arbeidsforholdService.hentArbeidsforhold(fnr)

        assertThat(arbeidsforholdList).isNotNull
        val arbeidsforhold = arbeidsforholdList!![0]
        assertThat(arbeidsforhold.arbeidsgivernavn).isEqualTo(orgNavn)
        assertThat(arbeidsforhold.harFastStilling).isTrue
        assertThat(arbeidsforhold.fastStillingsprosent).isEqualTo(100L)
        assertThat(arbeidsforhold.orgnr).isEqualTo(orgnr)
        assertThat(arbeidsforhold.fom).isEqualTo(fom.format(DateTimeFormatter.ISO_LOCAL_DATE))
        assertThat(arbeidsforhold.tom).isEqualTo(tom.format(DateTimeFormatter.ISO_LOCAL_DATE))
    }

    @Test
    internal fun skalSetteArbeidsgivernavnTilOrgnrHvisArbeidsgiverErOrganisasjon() {
        every { aaregClient.finnArbeidsforholdForArbeidstaker(fnr) } returns listOf(createArbeidsforhold(false, fom, tom))

        val arbeidsforholdList = arbeidsforholdService.hentArbeidsforhold(fnr)
        val arbeidsforhold = arbeidsforholdList!![0]

        assertThat(arbeidsforhold.arbeidsgivernavn).isEqualTo("Privatperson")
        assertThat(arbeidsforhold.orgnr).isNull()
    }

    @Test
    internal fun skalAddereStillingsprosentFraArbeidsavtaler() {
        every { aaregClient.finnArbeidsforholdForArbeidstaker(fnr) } returns listOf(createArbeidsforholdMedFlereArbeidsavtaler(12.3, 45.6))
        every { organisasjonService.hentOrgNavn(any()) } returns orgNavn

        val arbeidsforholdList = arbeidsforholdService.hentArbeidsforhold(fnr)
        val arbeidsforhold = arbeidsforholdList!![0]

        // desimaler strippes fra double til long
        assertThat(arbeidsforhold.fastStillingsprosent).isEqualTo(57L)
    }

    @Test
    internal fun ansettelsesperiodeTomErNull() {
        every { aaregClient.finnArbeidsforholdForArbeidstaker(fnr) } returns listOf(createArbeidsforhold(true, fom, null))
        every { organisasjonService.hentOrgNavn(any()) } returns orgNavn

        val arbeidsforholdList = arbeidsforholdService.hentArbeidsforhold(fnr)
        val arbeidsforhold = arbeidsforholdList!![0]

        assertThat(arbeidsforhold.arbeidsgivernavn).isEqualTo(orgNavn)
        assertThat(arbeidsforhold.harFastStilling).isTrue
        assertThat(arbeidsforhold.fastStillingsprosent).isEqualTo(100L)
        assertThat(arbeidsforhold.orgnr).isEqualTo(orgnr)
        assertThat(arbeidsforhold.fom).isEqualTo(fom.format(DateTimeFormatter.ISO_LOCAL_DATE))
        assertThat(arbeidsforhold.tom).isNull()
    }

    private fun createArbeidsforhold(erArbeidsgiverOrganisasjon: Boolean, fom: LocalDate, tom: LocalDate?): ArbeidsforholdDto {
        val ansettelsesperiodeDto = AnsettelsesperiodeDto(PeriodeDto(fom, tom))
        val arbeidsavtaleDto = ArbeidsavtaleDto(100.0)
        return ArbeidsforholdDto(
            ansettelsesperiodeDto,
            listOf(arbeidsavtaleDto),
            "arbeidsforholdId",
            if (erArbeidsgiverOrganisasjon) arbeidsgiverOrganisasjon else arbeidsgiverPerson,
            arbeidstaker
        )
    }

    private fun createArbeidsforholdMedFlereArbeidsavtaler(vararg stillingsprosenter: Double): ArbeidsforholdDto {
        val ansettelsesperiodeDto = AnsettelsesperiodeDto(PeriodeDto(fom, tom))
        val arbeidsavtaler = stillingsprosenter.map { ArbeidsavtaleDto(it) }
        return ArbeidsforholdDto(
            ansettelsesperiodeDto,
            arbeidsavtaler,
            "arbeidsforholdId",
            arbeidsgiverOrganisasjon,
            arbeidstaker
        )
    }
}
